core/internal/cltest/cltest.go

```diff
@@ -1452,6 +1426,30 @@ func (b *Blocks) NewHead(number uint64) *evmtypes.Head {
 	return head
 }
 
+// Slice returns a slice of heads from number i to j. Set j to -1 for all remaining.
+func (b *Blocks) Slice(i, j int) (heads []*evmtypes.Head) {
+	b.t.Logf("Slicing heads from %v to %v...", i, j)
+
+	if j > 0 {
+		l := j - i
+		if l > len(b.Heads) {
+			b.t.Fatalf("invalid configuration: too few blocks %d for range length %d", len(b.Heads), l)
+		}
+		heads = make([]*evmtypes.Head, 0, l)
+	}
+	for n := i; j < 0 || n < j; n++ {
+		h, ok := b.Heads[int64(n)]
+		if !ok {
+			if j < 0 {
+				break // done
+			}
+			b.t.Fatalf("invalid configuration: block %d not found", n)
+		}
+		heads = append(heads, h)
+	}
+	return
+}
+
```

Comment:
Totally up to you, but I was playing around with this and think this might be a little cleaner. 

One difference with this suggestion is that error handling is left to the consumer of `Slice`, which imho I think is better as `Slice` seems to be responsible for validating and performing the slice, but the consumer who calls Slice should handle any errors related to issues with the blocks needed for the slice. What do you think?

```suggestion
// Slice returns a slice of heads from i to j.
// Set j to -1 for all remaining heads.
// Returns an error if the slice cannot be created.
func (b *Blocks) Slice(i, j int) ([]*evmtypes.Head, error) {
    b.t.Logf("Slicing heads from %v to %v...", i, j)

    if err := b.validateIndices(i, j); err != nil {
        return nil, err
    }

    return b.createHeadSlice(i, j), nil
}

// validateIndices checks if the provided indices are valid for slicing.
func (b *Blocks) validateIndices(i, j int) error {
    if j > i {
        sliceLength := j - i
        if sliceLength > len(b.Heads) {
            return fmt.Errorf("invalid configuration: too few blocks %d for range length %d", len(b.Heads), sliceLength)
        }
    }
    return nil
}

// createHeadSlice creates a slice of heads based on the provided indices.
func (b *Blocks) createHeadSlice(i, j int) []*evmtypes.Head {
    var heads []*evmtypes.Head
    if j > i {
        heads = make([]*evmtypes.Head, 0, j-i)
    }

    for index := i; j == -1 || index < j; index++ {
        head, exists := b.Heads[int64(index)]
        if !exists {
            if j == -1 {
                break // done
            }
            return nil // Assuming that an error will be returned if a block is not found
        }
        heads = append(heads, head)
    }

    return heads
}

```

Reply:
I think this structure reads easier, but for error handling I was actually thinking that a `panic` would be appropriate, like with a real slice or `package reflect` slice, but that since we are in test code it would be less disruptive to just kill the test instead of the whole suite.

Reply:
Ah, got it. LGTM

Reply:
updated